### PR TITLE
Implements "value-in-payable" rule

### PR DESCRIPTION
**Pull request by mushketyk**
Implements one of the rules proposed in https://github.com/duaraghav8/Solium/issues/44

### CODE REVIEW

File: config/solium.json [hunk 1]
```diff
@@ -120,6 +120,13 @@
 			"description": "Ensure that all strings use only 1 style - either double quotes or single quotes."
 		},
 
+		"value-in-payable": {
+			"enabled": true,
+			"recommended": true,
+			"type": "error",
+			"description": "Ensure that all strings use only 1 style - either double quotes or single quotes."
```
Comment (duaraghav8) [inline config/solium.json]: I think you forgot to change the rule's description here (happens with me a lot!)

File: lib/rules/value-in-payable.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        docs: {
+            recommended: true,
+            type: "error",
+            description: "Ensure 'msg.value' is only used in functions with the 'payable' modifier"
+        },
+
+        schema: []
+    },
+
+    create: function(context) {
```
Comment (duaraghav8) [inline lib/rules/value-in-payable.js]: please change it to `create(context) {`. + I also encourage you to use as many ES6 features as you'd like (I'm anyway moving the old solium code into es6 these days)

File: lib/rules/value-in-payable.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            currentFunctionDeclaration = emitted.node;
+
+        }
+
+        function inspectMemberExpression(emitted) {
+            if (emitted.exit) {
+                return;
+            }
+
+            let node = emitted.node;
```
Comment (duaraghav8) [inline lib/rules/value-in-payable.js]: please change this to `const { node } = emitted;`

File: lib/rules/value-in-payable.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+            if (_isMsgValueAccess(node) && _isInFunction() && !_isInPayableFunction()) {
+                context.report({
+                    node: emitted.node,
+                    message: "'msg.value' should only be used in a payable function"
```
Comment (duaraghav8) [inline lib/rules/value-in-payable.js]: This message could be a bit confusing. Please change it to `Function must be declared payable if using 'msg.value'`

File: lib/rules/value-in-payable.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                });
+            }
+        }
+
+        function _isMsgValueAccess(node) {
+            return node.object.name === "msg" && node.property.name == "value";
```
Comment (duaraghav8) [inline lib/rules/value-in-payable.js]: By convention, solium uses triple equals. So please change it to `node.property.name === "value";` (unless you have a specific reason to use double equals).

File: lib/rules/value-in-payable.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        function _isInFunction() {
+            return currentFunctionDeclaration != null;
```
Comment (duaraghav8) [inline lib/rules/value-in-payable.js]: please use `!==` instead.

**Comment by duaraghav8** [comment]
Thanks @mushketyk! I've requested a few minor changes, everything else looks good.

Once you've made those changes, please redirect this PR to `develop` branch (don't worry about build failures, I'll resolve them. develop is slightly ahead of master so you can expect build fails)

**Comment by mushketyk** [comment]
Hi @duaraghav8 

Thank you for the detailed review! I've updated the code and changed the merge target to the `develop` branch.

**Pull request merged**